### PR TITLE
Create a new PartialObjectMetadata focused client and switch GC to use it

### PR DESCRIPTION
**Pull request by smarterclayton**
As part of kubernetes/enhancements#929, we want to use PartialObjectMetadata from at least one client internally.

Create a new variant of the dynamic client (still WIP for code structure) that works with PartialObjectMetadata and use it in the garbage collector.

/kind feature

TODO:

* [x] Clean up the client and potentially move it to a new package `k8s.io/client-go/metadata` or similar
* [x] Add better client docs for the package
* [ ] Measure performance change to GC controller to validate
* [x] Complete integration test for client

```release-note
Introduce a new client that can deal with Kubernetes API objects generically by reading only their metadata and use it in the garbage collection controller to both simplify code and improve the performance of garbage collection.
```

### CODE REVIEW

**Comment by smarterclayton** [comment]
density CPU seemed lower than the average (0.20 vs the more normal ~0.4) but load was right in the middle

/test pull-kubernetes-e2e-gce-100-performance

**Comment by smarterclayton** [comment]
@wojtek-t how hard would it be to get controller manager profiles added to the density/load runs?

**Comment by wojtek-t** [comment]
> @wojtek-t how hard would it be to get controller manager profiles added to the density/load runs?

Shouldn't be hard - let me look into that tomorrow.

**Comment by wojtek-t** [comment]
> Shouldn't be hard - let me look into that tomorrow.

@smarterclayton Actually, I sent out https://github.com/kubernetes/perf-tests/pull/477 for review, which will start gathering CM profiles too.
The bad news is that apparently the currently existing profiles are useless, because there is exactly one profile gathered after test finishes - for cpu profile it doesn't make any sense.

I need to run now, but I will try to fix that tomorrow too.

**Comment by smarterclayton** [comment]
OK, great.

**Comment by wojtek-t** [comment]
@smarterclayton - it took some time, but thethe profiles are there now and they are really being gathered periodically. so should be fine now.

File: staging/src/k8s.io/client-go/metadata/interface.go [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+type Interface interface {
+	Resource(resource schema.GroupVersionResource) Getter
+}
+
+// ResourceInterface contains the set of methods that may be invoked on objects by their metadata.
+// Update is not supported by the server, but Patch can be used for the actions Update would handle.
+type ResourceInterface interface {
+	Delete(name string, options *metav1.DeleteOptions, subresources ...string) error
+	DeleteCollection(options *metav1.DeleteOptions, listOptions metav1.ListOptions) error
+	Get(name string, options metav1.GetOptions, subresources ...string) (*metav1beta1.PartialObjectMetadata, error)
```
Comment (liggitt) [inline staging/src/k8s.io/client-go/metadata/interface.go]: Hrmm, not thrilled about baking beta types into a signature in an unversioned package

File: staging/src/k8s.io/client-go/metadata/metadata.go [hunk 2]
```diff
@@ -0,0 +1,285 @@
...
+
+// NewForConfig creates a new metadata client that can retrieve object
+// metadata details about any Kubernetes object (core, aggregated, or custom
+// resource based) in the form of PartialObjectMetadata objects, or returns
+// an error.
+func NewForConfig(inConfig *rest.Config) (Interface, error) {
+	config := ConfigFor(inConfig)
+	// for serializing the options
+	config.GroupVersion = &schema.GroupVersion{}
+	config.APIPath = "/if-you-see-this-search-for-the-break"
```
Comment (liggitt) [inline staging/src/k8s.io/client-go/metadata/metadata.go]: I love how this copy/paste breaks the usefulness of the original

File: staging/src/k8s.io/client-go/metadata/metadata.go [hunk 3]
```diff
@@ -0,0 +1,285 @@
...
+		Do()
+	return result.Error()
+}
+
+func (c *client) Get(name string, opts metav1.GetOptions, subresources ...string) (*metav1beta1.PartialObjectMetadata, error) {
+	if len(name) == 0 {
+		return nil, fmt.Errorf("name is required")
+	}
+	result := c.client.client.Get().AbsPath(append(c.makeURLSegments(name), subresources...)...).
+		SetHeader("Accept", "application/vnd.kubernetes.protobuf;as=PartialObjectMetadata;g=meta.k8s.io;v=v1beta1,application/json;as=PartialObjectMetadata;g=meta.k8s.io;v=v1beta1,application/json").
```
Comment (liggitt) [inline staging/src/k8s.io/client-go/metadata/metadata.go]: what's the application/json fallback for? status errors only?

**Comment by liggitt** [comment]
metadata as a package is a tad vague, but it maps ok, given what the client is

my only real concern is the use of concrete beta types in signatures in an unversioned package

File: staging/src/k8s.io/client-go/metadata/metadata.go (hunk 3 above)
Comment (smarterclayton) [reply staging/src/k8s.io/client-go/metadata/metadata.go]: Older servers

File: staging/src/k8s.io/client-go/metadata/metadata.go (hunk 2 above)
Comment (smarterclayton) [reply staging/src/k8s.io/client-go/metadata/metadata.go]: Yeah

Comment (smarterclayton) [reply staging/src/k8s.io/client-go/metadata/metadata.go]: Will change

Comment (smarterclayton) [reply staging/src/k8s.io/client-go/metadata/metadata.go]: Actually, has this ever actually caught anything?

File: staging/src/k8s.io/client-go/metadata/interface.go (hunk 1 above)
Comment (smarterclayton) [reply staging/src/k8s.io/client-go/metadata/interface.go]: I can easily move it to GA.  Without internal conversion types, don't have many options.

Also, breaking client APIs is something we do all the time...

**Comment by smarterclayton** [comment]
Actually, re: skew, we need the ListMetadata protobuf fix in 1.14 if we are going to say (1.15 controllers have to work against 1.14 apiservers).

Wait, why do controllers have to work against older apiservers?

**Comment by liggitt** [comment]
> Wait, why do controllers have to work against older apiservers?

I was thinking of controller -> LB -> HA apiservers, but looks like you're right: https://kubernetes.io/docs/setup/version-skew-policy/#kube-controller-manager-kube-scheduler-and-cloud-controller-manager

I do want to understand what would happen with a 1.15 controller -> 1.14 apiserver with this PR as is

**Comment by smarterclayton** [comment]
List watches won't work against 1.14 apiservers.

**Comment by smarterclayton** [comment]
What is wrong with the jobs - @kubernetes/sig-testing I've seen this after multiple force pushes, do we have a bad cached ref somewhere in the job infra?

```
I0418 14:57:12.904] process 103 exited with code 0 after 0.0m
32
I0418 14:57:12.905] Call: git merge --no-ff -m 'Merge +refs/pull/76054/head:refs/pr/76054' 67dac6d0170bfc87a419cb7349a3748cfc0f24f3
33
W0418 14:57:12.939] merge: 67dac6d0170bfc87a419cb7349a3748cfc0f24f3 - not something we can merge
34
E0418 14:57:12.942] Command failed
```

**Comment by stevekuznetsov** [comment]
@smarterclayton `67dac6d0170bfc87a419cb7349a3748cfc0f24f3` is your latest commit, so the fact that we're testing it is not bad. This is a failure mode we've recently noticed and we thing it might be an issue with GitHub fetch. /cc @cjwagner

**Comment by stevekuznetsov** [comment]
(Specifically a lag in GitHub's `refs/pull/76054/head` refspec pointing to the right thing)

**Comment by wojtek-t** [comment]
@stevekuznetsov - I'm facing that quite frequently recently, e.g.: https://github.com/kubernetes/kubernetes/pull/76702

**Comment by liggitt** [comment]
> List watches won't work against 1.14 apiservers.

that's a pretty unfortunate failure mode... is it worth it over delaying the controller change by a release?

File: staging/src/k8s.io/client-go/metadata/metadata.go [hunk 4]
```diff
@@ -0,0 +1,285 @@
...
+		SpecificallyVersionedParams(&opts, dynamicParameterCodec, versionV1).
+		Do()
+	if err := result.Error(); err != nil {
+		return nil, err
+	}
+	obj, err := result.Get()
+	if runtime.IsNotRegisteredError(err) {
```
Comment (caesarxuchao) [inline staging/src/k8s.io/client-go/metadata/metadata.go]: When will this case happen?

Comment (caesarxuchao) [reply staging/src/k8s.io/client-go/metadata/metadata.go]: `klog.Infof("Error: %#v", err)` as in the LIST case?

File: staging/src/k8s.io/client-go/metadata/metadata.go [hunk 5]
```diff
@@ -0,0 +1,285 @@
...
+		klog.Infof("Error: %#v", err)
+		rawBytes, err := result.Raw()
+		if err != nil {
+			return nil, err
+		}
+		var partial metav1beta1.PartialObjectMetadataList
+		if err := json.Unmarshal(rawBytes, &partial); err != nil {
+			return nil, fmt.Errorf("unable to decode returned object as PartialObjectMetadataList: %v", err)
+		}
```
Comment (caesarxuchao) [inline staging/src/k8s.io/client-go/metadata/metadata.go]: Shall we check if `partial` is empty, like in the `Get` method?

**Review by caesarxuchao — COMMENTED** [review]
I just took one pass and left some initial questions. I'll take another pass tomorrow.

**Comment by smarterclayton** [comment]
> that's a pretty unfortunate failure mode... is it worth it over delaying the controller change by a release?

We can survive landing this in 1.16.  Question being do we delay the API reaching v1 for it then?

**Comment by liggitt** [comment]
> Question being do we delay the API reaching v1 for it then?

In the absence of skew-compatible client/server APIs (e.g. REST API promotions), we've rolled out the server a release before switching the clients

**Comment by smarterclayton** [comment]
I am unable to parse your wisdom - was that "let's try for v1 given input we have in 1.15" or "wait for v1".  If we put v1 in we can keep the dynamic client set only to v1

**Comment by liggitt** [comment]
we haven't delayed promoting an API to v1 in the past, just because we had to wait a release to switch clients to it

**Comment by liggitt** [comment]
the partial object and table types have been in v1beta1 for several releases with use, which gives me confidence to promote them, and switch clients when feasible

**Comment by smarterclayton** [comment]
ok.  I'll spin that up, get that merged, mark this as hold for 1.16, and potentially split the dynamic client out and merge that first so we have that in core for testing with namespace and others.

**Comment by smarterclayton** [comment]
x-ref v1 api https://github.com/kubernetes/kubernetes/pull/77136

**Comment by smarterclayton** [comment]
Split into https://github.com/kubernetes/kubernetes/pull/77819 and https://github.com/kubernetes/kubernetes/pull/77817, the GC bit will be moved to 1.16, the API is v1 in 1.15 (if we can land #77187).

**Pull request closed without merging**